data/abi/abi_encode.go

```diff
@@ -20,214 +20,249 @@ import (
+		return nil, fmt.Errorf("cannot infer type for encoding")
 	}
 }
 
-// compressMultipleBool compress consecutive bool values into a byte in ABI tuple/array value.
-func compressMultipleBool(valueList []Value) (uint8, error) {
-	var res uint8 = 0
-	if len(valueList) > 8 {
-		return 0, fmt.Errorf("value list passed in should be no greater than length 8")
+// encodeInt encodes int-alike golang values to bytes, following ABI encoding rules
+func encodeInt(intValue interface{}, bitSize uint16) ([]byte, error) {
+	var bigInt big.Int
+
+	switch intValue := intValue.(type) {
+	case int8:
+		if intValue < 0 {
+			return nil, fmt.Errorf("passed in int value should be non negative")
+		}
+		bigInt = *new(big.Int).SetUint64(uint64(intValue))
+	case uint8:
+		bigInt = *new(big.Int).SetUint64(uint64(intValue))
+	case int16:
+		if intValue < 0 {
+			return nil, fmt.Errorf("passed in int value should be non negative")
+		}
+		bigInt = *new(big.Int).SetUint64(uint64(intValue))
+	case uint16:
+		bigInt = *new(big.Int).SetUint64(uint64(intValue))
+	case int32:
+		if intValue < 0 {
+			return nil, fmt.Errorf("passed in int value should be non negative")
+		}
+		bigInt = *new(big.Int).SetUint64(uint64(intValue))
+	case uint32:
+		bigInt = *new(big.Int).SetUint64(uint64(intValue))
+	case int64:
+		if intValue < 0 {
+			return nil, fmt.Errorf("passed in int value should be non negative")
+		}
+		bigInt = *new(big.Int).SetUint64(uint64(intValue))
+	case uint64:
+		bigInt = *new(big.Int).SetUint64(intValue)
+	case uint:
+		bigInt = *new(big.Int).SetUint64(uint64(intValue))
+	case int:
+		if intValue < 0 {
+			return nil, fmt.Errorf("passed in int value should be non negative")
+		}
+		bigInt = *new(big.Int).SetUint64(uint64(intValue))
+	case *big.Int:
+		bigInt = *new(big.Int).Set(intValue)
+	default:
+		return nil, fmt.Errorf("cannot infer go type for uint encode")
 	}
-	for i := 0; i < len(valueList); i++ {
-		if valueList[i].ABIType.abiTypeID != Bool {
-			return 0, fmt.Errorf("bool type not matching in compressMultipleBool")
-		}
-		boolVal, err := valueList[i].GetBool()
-		if err != nil {
-			return 0, err
-		}
-		if boolVal {
-			res |= 1 << uint(7-i)
-		}
+
+	bytes := bigInt.Bytes()
+	if len(bytes) > int(bitSize/8) {
+		return nil, fmt.Errorf("input value bit size %d > abi type bit size %d", len(bytes)*8, bitSize)
 	}
-	return res, nil
+
+	zeroPadding := make([]byte, bitSize/8-uint16(len(bytes)))
+	buffer := append(zeroPadding, bytes...)
+	return buffer, nil
 }
 
-// tupleEncoding encodes an ABI value of tuple type into an ABI encoded byte string.
-func tupleEncoding(v Value) ([]byte, error) {
-	if v.ABIType.abiTypeID != Tuple {
-		return []byte{}, fmt.Errorf("type not supported in tupleEncoding")
+// inferToSlice infers an interface element to a slice of interface{}, returns error if it cannot infer successfully
+func inferToSlice(value interface{}) ([]interface{}, error) {
+	reflectVal := reflect.ValueOf(value)
+	if reflectVal.Kind() != reflect.Slice {
```

Comment:
Maybe I understood it wrong, but I think if you are passing a slice, then the slice is automatically casted into an interface...?

Reply:
I mean it would be nice for a go array like `[10]bool` to be a valid input as well. This type has kind `reflect.Array`, but other than that, it's pretty much the same as a slice for our purposes.